docs/core/extensions/snippets/logging/library-authors/DiExampleService.cs

```diff
@@ -0,0 +1,19 @@
+﻿using Microsoft.Extensions.Logging;
+
+namespace Logging.LibraryAuthors;
+
+public class DiExampleService
+{
+    private readonly ILogger<DiExampleService> _logger;
+
+    public DiExampleService(ILogger<DiExampleService> logger) =>
+        _logger = logger;
+
+    public void DoSomething()
+    {
+        if (_logger.IsEnabled(LogLevel.Information))
```

Comment:
I would only recommend adding these IsEnabled() checks if the code needs to do some non-trivial work to create the arguments to the logging call. Often it doesn't in which case I suspect developers are making their code more complicated for no benefit.

Reply:
'Non-trivial' will depend on the developer's performance goals. At the high-performance end, some developers might consider a single object allocation as non-trivial overhead. An example of that would be calling one of the LogXXX() extension methods that takes a `params` array. For example:
```C#
_logger.LogInformation("Sold {count} of {product}", 2, "cookies");
```

That example boxes the integer 2 into an object, then creates an object[] to hold the boxed `2` and the "cookies" string. A total of two object allocations. The source generated logging doesn't have those extra allocations in it btw, only the LogXXX extension methods.

For other users a couple of object allocations might be no big deal, but maybe they will be calling a method that has larger perf overheads. For example:

```C#
[LoggerMessage(Message="Sold {count} of {product}")]
static void ProductSold(ILogger logger, int count, string product);

ProductSold(logger, 2, product.GetFriendlyProductDescription());
```

`GetFriendlyProductDescription()` might do some formatting or load a value from a file or query a database. Whatever it is, if it isn't a perf cost the user wants to pay for when the logging is disabled then they can use the IsEnabled() check first to avoid it.

Reply:
The 2nd example might be better than the first. If you are a developer who cares about high performance enough to consider adding the IsEnabled() check to the first example, you'd be better off making the change to source generated logging instead.

Reply:
I see what you're saying, and that makes sense, but the code you suggested:

```cs
[LoggerMessage(Message="Sold {count} of {product}")]
static void ProductSold(ILogger logger, int count, string product);

ProductSold(logger, 2, product.GetFriendlyProductDescription());
```

Will always evaluate the `product.GetFriendlyProductDescription()` as it's a `string` parameter of the log method — regardless of whether it's actually logged (`logLevel` enabled), that method is still invoked to provide the argument to the log method. The source generator isn't actually helping here as the overhead of the call will still exist, right?